marlowe-dashboard-client/src/InputField/State.purs

```diff
@@ -39,6 +54,33 @@ handleAction (SetValue value) =
     $ set _value value
     <<< set _pristine false
 
+-- This handler is used for number inputs instead of `SetValue`. Since `getBigIntegerValue` and
+-- `humanizeValue` are essentially inverses of each other, you might think it wouldn't do anything
+-- differently from what plain old SetValue does. But the difference is that this version sanitizes
+-- the user input, replacing empty strings with "0" and enforcing the right number of decimal
+-- places. Also there's an extra wrinkle. If the numeric value hasn't actually changed, then it
+-- seems Halogen doesn't bother updating the DOM. A sensible optimisation in general, but in this
+-- case we need to force the update, because the string value can change without the numeric value
+-- value changing, and we still want to update in that case. For example, if you add extra integers
+-- to the end of the string (beyond the number of decimal places allowed by the format), then those
+-- are discarded when the string is parsed - but before I added a workaround for this they would
+-- still show up in the DOM. To fix this, we manually set a different value before setting the new
+-- value. This forces the update to the DOM, and happens so fast that you don't notice it.
+handleAction (SetFormattedValue numberFormat value) = do
+  currentValue <- use _value
+  let
+    decimals = case numberFormat of
+      DefaultFormat -> zero
+      DecimalFormat d _ -> d
+
+    bigIntegerValue = getBigIntegerValue decimals value
+
+    newValue = humanizeValue decimals bigIntegerValue
+
+    forcedDifferentValue = humanizeValue decimals $ bigIntegerValue + (BigInteger.fromInt 1)
```

Comment:
I'm not sure I understand this part. I get the fact that in some cases we need to force halogen for a change, but isn't this modifying the value?

Reply:
I set the modified value and then immediately set it back to the right value. Do you know of a less hacky way to force the change?